client_test.go

```diff
@@ -0,0 +1,65 @@
+package metricd
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+)
+
+// Should be run in a separate goroutine. Runs a grpc server at the
+// given port after sleeping for 3 seconds. The listener will be passed
+// to listenerChan, or listenerChan will be closed if there was an error.
+func runSlowListeningServer(t *testing.T, port int, listenerChan chan net.Listener) {
+	time.Sleep(3 * time.Second)
+	server := grpc.NewServer()
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		fmt.Println(err)
+		close(listenerChan)
+		return
+	}
+	// Let the main goroutine know about the listener so they can close it
+	listenerChan <- l
+	server.Serve(l)
+}
+
+func TestClientConnect(t *testing.T) {
+	// Set up a dummy grpc server that takes a few seconds to
+	// start listening
+	const port = 6789
+	listenerChan := make(chan net.Listener, 1)
+	go runSlowListeningServer(t, port, listenerChan)
+	defer func() {
+		l, success := <-listenerChan
+		if success {
+			l.Close()
+		}
+	}()
+
+	// Create a client
+	url := url.URL{
+		Scheme: "http",
+		Host:   fmt.Sprintf("localhost:%d", port),
+	}
+	client, err := NewClient(&url)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	// Try connecting
+	rc := client.(*realClient)
+	err = rc.tryConnect()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if rc.conn.State() != grpc.Ready {
```

Comment:
This is technically racy, for the same reason that the code that you are replacing is racy: by the time you perform this check, the connection may have gone out of ready state due to a transient network disruption.

OTOH, since this connection runs on the loopback interface, and the amount of load going over it is trivial, it is unlikely for this to happen in practice. But on a heavily loaded CI machine, this test could potentially be flaky. I suggest adding a comment acknowledging this possibility.